src/QueryDescription/test.js

```diff
@@ -267,32 +274,108 @@ describe('buildQueryDescription', () => {
+      ]),
+    ])
+    expect(query).toEqual({
+      where: [
+        {
+          type: 'on',
+          table: 'projects',
+          conditions: [
+            { type: 'where', left: 'foo', comparison: { operator: 'eq', right: { value: 'bar' } } },
+            {
+              type: 'or',
+              conditions: [
+                {
+                  type: 'where',
+                  left: 'bar',
+                  comparison: { operator: 'eq', right: { value: 'baz' } },
+                },
+                {
+                  type: 'where',
+                  left: 'bla',
+                  comparison: { operator: 'eq', right: { value: 'boop' } },
+                },
+              ],
+            },
+          ],
+        },
+      ],
+      joinTables: ['projects'],
+      nestedJoinTables: [],
+      sortBy: [],
+      take: null,
+      skip: null,
+    })
+  })
+  it(`supports deep nesting Q.on inside Q.on`, () => {
     const query = Q.buildQueryDescription([
-      Q.experimentalJoinTables(['projects']),
       Q.experimentalNestedJoin('projects', 'teams'),
-      Q.on('projects', Q.on('teams', 'foo', 'bar')),
+      Q.experimentalNestedJoin('teams', 'organizations'),
+      Q.on('projects', Q.on('teams', Q.on('organizations', 'foo', 'bar'))),
     ])
     expect(query).toEqual({
       where: [
         {
           type: 'on',
           table: 'projects',
-          nested: {
-            type: 'on',
-            table: 'teams',
-            left: 'foo',
-            comparison: { operator: 'eq', right: { value: 'bar' } },
-          },
+          conditions: [
+            {
+              type: 'on',
+              table: 'teams',
+              conditions: [
+                {
+                  type: 'on',
+                  table: 'organizations',
+                  conditions: [
+                    {
+                      type: 'where',
+                      left: 'foo',
+                      comparison: { operator: 'eq', right: { value: 'bar' } },
+                    },
+                  ],
+                },
+              ],
+            },
+          ],
         },
       ],
       joinTables: ['projects'],
-      nestedJoinTables: [{ from: 'projects', to: 'teams' }],
+      nestedJoinTables: [{ from: 'projects', to: 'teams' }, { from: 'teams', to: 'organizations' }],
       sortBy: [],
       take: null,
       skip: null,
     })
   })
+  it(`supports Q.on shortcut syntaxes`, () => {
+    const expected = Q.on('projects', [Q.where('foo', Q.eq('bar'))])
+    expect(Q.on('projects', 'foo', 'bar')).toEqual(expected)
+    expect(Q.on('projects', 'foo', Q.eq('bar'))).toEqual(expected)
+    expect(Q.on('projects', Q.where('foo', 'bar'))).toEqual(expected)
+    expect(Q.on('projects', Q.and(Q.where('foo', 'bar'), Q.where('bar', 'baz')))).toEqual(
+      Q.on('projects', [Q.where('foo', 'bar'), Q.where('bar', 'baz')]),
+    )
+  })
+  it(`compresses top-level Q.ons into a single nested Q.on`, () => {
+    expect(
```

Comment:
// Multiple separate Q.ons is a legacy syntax producing suboptimal query code unless
  // special cases are used. Here, we're special casing only top-level Q.ons to avoid regressions
  // but it's not recommended for new code
  // TODO: Remove this special case